lib/iris/analysis/__init__.py

```diff
@@ -1152,7 +1200,31 @@ def inner_stat(array, axis=-1, mdtol=None, **kwargs):
     return inner_stat
 
 
-def _percentile(data, axis, percent, fast_percentile_method=False, **kwargs):
+def _calc_percentile(data, percent, fast_percentile_method=False, **kwargs):
+    """
+    Calculate percentiles along the trailing axis of a 1D or 2D array.
+
+    """
+    if fast_percentile_method:
+        msg = "Cannot use fast np.percentile method with masked array."
+        if ma.is_masked(data):
+            raise TypeError(msg)
+        result = np.percentile(data, percent, axis=-1)
+        result = result.T
+    else:
+        quantiles = percent / 100.0
+        result = scipy.stats.mstats.mquantiles(
```

Comment:
I saw your example in https://github.com/SciTools/iris/issues/4039 where you fill the masked data with NaN, so you can use a nanmedian function. I wonder if the same approach would work here, as numpy does have a nanpercentile function.

Reply:
Maybe that's beyond the scope of this PR, though. I just tried to understand what's going on.

Reply:
The idea of using `np.nanpercentile` for the masked arrays is attractive, since it has the same calling sequence as `np.percentile` (used in our "fast percentile method") so we could simplify our API.  However the scipy `mquantiles` function does seem to be a lot more flexible and we would be losing functionality for users who care about particular interpolation methods.  So I think we'd need to retain it somehow.

As you say, I think it's out of scope for this PR, but useful to add to a wider discussion about how to organise the various percentile options (#3294).